nixos/modules/virtualisation/digital-ocean-config.nix

```diff
@@ -0,0 +1,206 @@
+{ config, pkgs, lib, modulesPath, ... }:
+with lib;
+{
+  imports = [
+    (modulesPath + "/profiles/qemu-guest.nix")
+    (modulesPath + "/virtualisation/digital-ocean-init.nix")
+  ];
+  options.virtualisation.digitalOcean = with types; {
+    setRootPassword = mkOption {
+      type = bool;
+      default = true;
+      example = true;
+      description = "Whether to set the root password from the Digital Ocean metadata";
+    };
+    setSshKeys = mkOption {
+      type = bool;
+      default = true;
+      example = true;
+      description = "Whether to fetch ssh keys from Digital Ocean";
+    };
+    seedEntropy = mkOption {
+      type = bool;
+      default = true;
+      example = true;
+      description = "Whether to run the kernel RNG entropy seeding script from the Digital Ocean vendor data";
+    };
+  };
+  config =
+    let
+      cfg = config.virtualisation.digitalOcean;
+      hostName = config.networking.hostName;
+    in mkMerge [{
+      fileSystems."/" = {
+        device = "/dev/disk/by-label/nixos";
+        autoResize = true;
+        fsType = "ext4";
+      };
+      boot = {
+        growPartition = true;
+        kernelParams = [ "console=ttyS0" "panic=1" "boot.panic_on_fail" ];
+        initrd.kernelModules = [ "virtio_scsi" ];
+        kernelModules = [ "virtio_pci" "virtio_net" ];
+        loader = {
+          grub.device = "/dev/vda";
+          timeout = 0;
+          grub.configurationLimit = 0;
+        };
+      };
+      services.openssh = {
+        enable = mkDefault true;
+        passwordAuthentication = mkDefault false;
+      };
+      networking = {
+        firewall.allowedTCPPorts = [ 22 ];
+        hostName = mkDefault ""; # use Digital Ocean metadata server
+      };
+
+      /* Check for and wait for the metadata server to become reachable.
+       * This serves as a dependency for all the other metadata services. */
+      systemd.services.digitalocean-metadata = {
+        path = [ pkgs.curl ];
+        description = "Get host metadata provided by Digitalocean";
+        script = ''
+          set -eu
+          DO_DELAY_ATTEMPTS=0
+          while ! curl -fsSLI http://169.254.169.254/metadata/v1.json > /dev/null; do
```

Comment:
I'm more worried that unless you explicitly firewall it, any process can retrieve the same data. Though, the root password service in particular I think needs some changes and should probably also be disabled by default.

> Also it would be nice to download all files in this unit and store them somewhere to be accessed by the others (in a location only readable by root).

I'd considered that but decided against it just because I didn't feel like deciding on a proper path for it or dealing with the corner cases:

1. The hostname can be changed while the server is running, so you can restart that service to update it. Making it depend on data from the first makes that dependency relationship (and the meaning of the systemd service states) a bit more complicated?
2. Each service was interested in a different endpoint, so downloading them all when most services never even run seemed like an unnecessary extra second of boot time.

Though, I believe the json blob contains everything... so each service can probably just use jq to pick out the pieces they need from that one file. I believe the hostname one is the only mutable piece of data so it could just continue to use the network to simplify things.

Reply:
Nope this is a Link-Local address so should be safe.

Reply:
"should be safe" doesn't make me feel safe! Can we check that it's a link-local address?

Reply:
It is by definition a link local address . Ipv4 mandates that the IP-range isn't next-hopped https://tools.ietf.org/html/rfc3927. Only people on the same physical link can talk over  the link-local ip range

Reply:
Oh nice! I feel safe now